src/main/java/ru/yandex/qatools/allure/cucumberjvm/AllureRunListener.java

```diff
@@ -49,45 +51,102 @@ public void testRunStarted(Description description) {
+     * @return {@link List<Description>} features description list
+     * @throws IllegalAccessException
+     */
+    private List<Description> findFeaturesLevel(List<Description> description)
+            throws IllegalAccessException {
+        if (description.isEmpty()) {
+            return new ArrayList<>();
+        }
+        Object readField = FieldUtils.readField(description.get(0), "fUniqueId", true);
+        if (readField instanceof Feature) {
+            return description;
+        } else {
+            return findFeaturesLevel(description.get(0).getChildren());
+        }
+
+    }
+
+    /**
+     * Find test classes level
+     *
+     * @param description {@link Description} Description to search where
+     * @return {@link List<Description>} test classes description list
+     * @throws IllegalAccessException
+     */
+    public List<Description> findTestClassesLevel(List<Description> description) throws IllegalAccessException {
+        if (description.isEmpty()) {
+            return new ArrayList<>();
+        }
+        Object possibleClass = FieldUtils.readField(description.get(0), "fUniqueId", true);
+        if (possibleClass instanceof String && !((String) possibleClass).isEmpty()) {
+            if (!description.get(0).getChildren().isEmpty()) {
+                Object possibleFeature = FieldUtils.readField(description.get(0).getChildren().get(0), "fUniqueId", true);
+                if (possibleFeature instanceof Feature) {
+                    return description;
+                } else {
+                    return findTestClassesLevel(description.get(0).getChildren());
+                }
+            } else {
+                //No scenarios in feature
+                return description;
+            }
+
+        } else {
+            return findTestClassesLevel(description.get(0).getChildren());
+        }
+
+    }
+
+    /**
      * Find feature and story for given scenario
      *
      * @param scenarioName
-     * @return array of {"<FEATURE_NAME>", "<STORY_NAME>"}
+     * @return {@link String[]} of ["<FEATURE_NAME>", "<STORY_NAME>"]s
      * @throws IllegalAccessException
      */
-    String[] findFeatureByScenarioName(String scenarioName) throws IllegalAccessException {
-        ArrayList<Description> features = parentDescription.getChildren().get(0).getChildren();
-        //Feature cycle
-        for (Description feature : features) {
-            //Story cycle
-            for (Description story : feature.getChildren()) {
-                Object scenarioType = FieldUtils.readField(story, "fUniqueId", true);
-
-                //Scenario
-                if (scenarioType instanceof Scenario) {
-                    if (story.getDisplayName().equals(scenarioName)) {
-                        return new String[]{feature.getDisplayName(), scenarioName};
-                    }
+    private String[] findFeatureByScenarioName(String scenarioName) throws IllegalAccessException {
```

Comment:
Is it really complicated. And the only I can do - the matrioshka. I`ve commented out all cycles to reduce complexity. And I think if I split it to several methods which nest each other, it will reduce readability of code.